### PR TITLE
Allow read-only file system in Kiali's pod

### LINKED ISSUE #3810: Modify manifests to use readOnlyRootFilesystem
**Is your feature request related to a problem? Please describe.**
It's a good security practice to use readOnlyRootFilesystems when deploying workloads. During kiali bootstrap, kiali tries to update the env.js and index.html (rootPath), which violate the readOnlyRootFilesystem condition.

**Describe the solution you'd like**
Update the documentation manifests to restrict readOnlyRootFilesystem.

**Describe alternatives you've considered**
I'm including a sample manifest that I'm using that uses an initContainer to copy the static files over to a memory volume.

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: kiali
spec:
  replicas: 1
  selector:
    matchLabels:
      app: kiali
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      name: kiali
      labels:
        app: kiali
      annotations:
        prometheus.io/scrape: "true"
        prometheus.io/port: "9090"
        kiali.io/runtimes: go,kiali
    spec:
      serviceAccountName: kiali
      initContainers:
        - name: copy
          securityContext:
            runAsUser: 1000
            runAsGroup: 1000
            runAsNonRoot: true
            allowPrivilegeEscalation: false
            readOnlyRootFilesystem: true
            privileged: false
            capabilities:
              drop:
                - ALL
          image: "quay.io/kiali/kiali:v1.31.0"
          command:
            - sh
            - -c
            - cp -r /opt/kiali/console/* /tmp/static
          volumeMounts:
            - mountPath: /tmp/static
              name: static-content
      containers:
        - image: "quay.io/kiali/kiali:v1.31.0"
          imagePullPolicy: Always
          name: kiali
          securityContext:
            runAsUser: 1000
            runAsGroup: 1000
            runAsNonRoot: true
            allowPrivilegeEscalation: false
            readOnlyRootFilesystem: true
            privileged: false
            capabilities:
              drop:
                - ALL
          command:
            - "/opt/kiali/kiali"
            - "-config"
            - "/kiali-configuration/config.yaml"
          ports:
            - name: api-port
              containerPort: 20001
            - name: http-metrics
              containerPort: 9090
          readinessProbe:
            httpGet:
              path: /kiali/healthz
              port: api-port
              scheme: HTTP
            initialDelaySeconds: 5
            periodSeconds: 30
          livenessProbe:
            httpGet:
              path: /kiali/healthz
              port: api-port
              scheme: HTTP
            initialDelaySeconds: 5
            periodSeconds: 30
          env:
            - name: ACTIVE_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
            - name: LOG_LEVEL
              value: "info"
            - name: LOG_FORMAT
              value: "text"
            - name: LOG_TIME_FIELD_FORMAT
              value: "2006-01-02T15:04:05Z07:00"
            - name: LOG_SAMPLER_RATE
              value: "1"
          volumeMounts:
            - name: kiali-configuration
              mountPath: "/kiali-configuration"
            - name: kiali-cert
              mountPath: "/kiali-cert"
            - name: kiali-secret
              mountPath: "/kiali-secret"
            - name: static-content
              mountPath: /opt/kiali/console
      volumes:
        - name: kiali-configuration
          configMap:
            name: kiali
        - name: kiali-cert
          secret:
            secretName: istio.kiali-service-account
            optional: true
        - name: kiali-secret
          secret:
            secretName: kiali
            optional: true
        - name: static-content
          emptyDir:
            medium: Memory
            sizeLimit: 50Mi
```

### PR DESCRIPTION
**Pull request by israel-hdez**
Specifically, stop writing changes to the index.html and env.js files, which come from the UI package.

These files are now generated on the back-end and served directly, rather than changing the files on the pod filesystem.

This allows using `readOnlyRootFilesystem: true` for the securityContext in the Kiali pod.

Related FAQ entry is removed in: https://github.com/kiali/kiali.io/pull/375

Fixes #3810

### CODE REVIEW

File: routing/router.go
Comment (jmazzitelli) [inline routing/router.go:142]: Shouldn't we return some kind of HTTP error status code and message here? It seems like a simple return will not provide the client any indication that a problem occurred. We should return some kind of 500 error here I think (writing it to "w")

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
@israel-hdez I didn't propose any implementation for the change I asked for - but I think we need to do something there in case of an error to let the user know something went wrong.

**Comment by jmazzitelli** [comment]
One thing I did not comment on in my review because I think it could be something we do later if we need such an enhancement for performance, but this new code does not cache those files in memory. Those files (both env.js and index.html) never change for the lifetime of the Kiali server, so why couldn't we build the files one time (say, on the first request), cache their contents, and then serve the cached contents thereafter?

I don't know how often the env.js file is requested (that isn't requested for every UI request right? Just when the UI is initialized/loaded?). I don't know how much we save if we cache, but it seems like such a low-hanging fruit/easy-win that I don't see any reason why we shouldn't build these just once and cache them.

**Comment by jmazzitelli** [comment]
Now that we do not require write access, we should lock down the pod security a bit more.

We should change these files:

* https://github.com/kiali/kiali-operator/blob/master/roles/default/kiali-deploy/templates/kubernetes/deployment.yaml
* https://github.com/kiali/kiali-operator/blob/master/roles/default/kiali-deploy/templates/openshift/deployment.yaml
* https://github.com/kiali/helm-charts/blob/master/kiali-server/templates/deployment.yaml

such that in the Kiali Deployment's `kiali` container, we add these constraints:

```
        securityContext:
          runAsUser: 1000
          runAsGroup: 1000
          runAsNonRoot: true
          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
          privileged: false
          capabilities:
            drop:
            - ALL
```

I will create a separate github issue to do this work, as it is not required for this PR and can be done separately.

File: routing/router.go
Comment (israel-hdez) [reply routing/router.go:143]: I added the message.
However, it broke some tests. I fixed them, but I don't like the fixes.

**Comment by israel-hdez** [comment]
> One thing I did not comment on in my review because I think it could be something we do later if we need such an enhancement for performance, but this new code does not cache those files in memory. Those files (both env.js and index.html)

Both files are loaded on each load/refresh.

For the `env.js` everything is generated at runtime from memory. It just creates a string from two configs. So I think performance impact should be negligible. I would say it's tie, because before this, it was a read from filesystem. Now, it's an in-memory process.

For the index.html, there is the penalty of an added search&replace operation on top of the usual filesytem read. TBH I think the performance impact if that added search&replaces operation should be negligible because this is an in-pod operation. Usually, we suffer from performance issues when we do a trip to the network (query the API Server, p8s, jaeger, etc.). So, I would like to try first without any cache (i.e. this time, I prefer to save memory rather than save CPU cycles).

**Comment by jmazzitelli** [comment]
> I would like to try first without any cache (i.e. this time, I prefer to save memory rather than save CPU cycles).

Actually, I meant cache it in the *server* (not the client). So the server would generate it once and only once. Store it in some go variable someplace.

But I agree we should wait to see how it goes. We can add that later if we need to.

File: routing/router_test.go
Comment (jmazzitelli) [inline routing/router_test.go:107]: is this even testing anything now? It should result in a successful request, but if we can't write to index.html because we don't have a console/index.html file, then this isn't testing what it was testing before, right?

File: routing/router_test.go
Comment (jmazzitelli) [inline routing/router_test.go:114]: same as my other comment

File: server/server_test.go
Comment (jmazzitelli) [inline server/server_test.go:78]: same as my other comment - this isn't really testing successfully that we are able to get to the custom root path.

File: server/server_test.go
Comment (jmazzitelli) [inline server/server_test.go:83]: same as my other comment - this isn't really testing successfully that we are able to get to the custom root path.

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
let's think about this some more and figure out how to get this to actually test what was being tested before (unless I am missing something with these changed tests).

**Comment by jmazzitelli** [comment]
@israel-hdez I got the tests to pass with this change:

```diff
diff --git a/routing/router_test.go b/routing/router_test.go
index f22d47d5..15975d58 100644
--- a/routing/router_test.go
+++ b/routing/router_test.go
@@ -4,6 +4,7 @@ import (
 	"io/ioutil"
 	"net/http"
 	"net/http/httptest"
+	"os"
 	"testing"
 
 	"github.com/gorilla/mux"
@@ -85,6 +86,12 @@ func TestRedirectWithSetWebRootKeepsParams(t *testing.T) {
 	oldConfig := config.Get()
 	defer config.Set(oldConfig)
 
+	oldWd, _ := os.Getwd()
+	defer os.Chdir(oldWd)
+	os.Chdir(os.TempDir())
+	os.MkdirAll("./console", 0777)
+	os.Create("./console/index.html")
+
 	conf := new(config.Config)
 	conf.Server.WebRoot = "/test"
 	config.Set(conf)
@@ -104,14 +111,14 @@ func TestRedirectWithSetWebRootKeepsParams(t *testing.T) {
 		t.Fatal(err)
 	}
 	body, _ := ioutil.ReadAll(resp.Body)
-	assert.Equal(t, 500, resp.StatusCode, "Response should not redirect")
+	assert.Equal(t, 200, resp.StatusCode, "Response should not redirect")
 
 	resp, err = client.Get(ts.URL + "/test/")
 	if err != nil {
 		t.Fatal(err)
 	}
 	body2, _ := ioutil.ReadAll(resp.Body)
-	assert.Equal(t, 500, resp.StatusCode, string(body2))
+	assert.Equal(t, 200, resp.StatusCode, string(body2))
 
 	assert.Equal(t, string(body), string(body2), "Response with and without the trailing slash on the webroot are not the same")
 }
diff --git a/server/server_test.go b/server/server_test.go
index 330b6b2b..3aab4a2c 100644
--- a/server/server_test.go
+++ b/server/server_test.go
@@ -30,6 +30,12 @@ const (
 var tmpDir = os.TempDir()
 
 func TestRootContextPath(t *testing.T) {
+	oldWd, _ := os.Getwd()
+	defer os.Chdir(oldWd)
+	os.Chdir(os.TempDir())
+	os.MkdirAll("./console", 0777)
+	os.Create("./console/index.html")
+
 	testPort, err := getFreePort(testHostname)
 	if err != nil {
 		t.Fatalf("Cannot get a free port to run tests on host [%v]", testHostname)
@@ -75,12 +81,12 @@ func TestRootContextPath(t *testing.T) {
 	checkHTTPReady(httpClient, serverURL)
 
 	// we should be able to get to our custom web root
-	if _, err = getRequestResults(t, httpClient, serverURL+testCustomRoot, noCredentials); err != nil && err.Error() != "Bad status: 500" {
+	if _, err = getRequestResults(t, httpClient, serverURL+testCustomRoot, noCredentials); err != nil {
 		t.Fatalf("Failed: Shouldn't have failed going to the web root: %v", err)
 	}
 
 	// we should be able to get to "/" root - this just forwards to our custom web root
-	if _, err = getRequestResults(t, httpClient, serverURL, noCredentials); err != nil && err.Error() != "Bad status: 500" {
+	if _, err = getRequestResults(t, httpClient, serverURL, noCredentials); err != nil {
 		t.Fatalf("Failed: Shouldn't have failed going to / root: %v", err)
 	}
```

File: routing/router_test.go
Comment (israel-hdez) [reply routing/router_test.go:107]: My thought is that this is testing that a redirection is *not* happening (or that's what I understand from the "Response should not redirect" string). 
Specifically, I *think* this is testing that this line is not being hit: https://github.com/kiali/kiali/blob/master/routing/router.go#L28. That line would return a `302` status code. So, my rationale was that as long as we don't get the `302` status code, the test should be green.

I had a similar rationale for `server_test.go`: as long as we get that `500` response, which is the expected within the test, I think it should be OK.

But I think mocking the index.html at runtime is fine. I'll do the changes.

**Comment by israel-hdez** [comment]
@jmazzitelli OK, I applied your suggestion to the tests.
Let's hope we won't see sporadic failures, because CI runs `make test-race` which may behave slightly different than `make test`.

**Comment by jmazzitelli** [comment]
linter failed - IMO, we need to skip the test code when lint is run

**Comment by israel-hdez** [comment]
@jmazzitelli OK, CI is fixed by assigning errs to the underscore :D

**Review by jmazzitelli — APPROVED** [review]
lint made this so much easier to read! :)

ltgm

**Comment by jmazzitelli** [comment]
Let's merge these three all together:
* this PR
* https://github.com/kiali/helm-charts/pull/76
* https://github.com/kiali/kiali-operator/pull/316

**Comment by lucasponce** [comment]
@israel-hdez @jmazzitelli please, why this work has been merged without QE in the review / approval ?

**Comment by jmazzitelli** [comment]
This work would have required QE to pull and review/test 3 separate PRs together as one. Figured it was easy enough for me and Edgar to test this stuff - it is easy enough to validate.

**Comment by lucasponce** [comment]
> This work would have required QE to pull and review/test 3 separate PRs together as one. Figured it was easy enough for me and Edgar to test this stuff - it is easy enough to validate.

These kind of changes, unless you have performed regression tests, are better that are in QE radar as well to validate that things look ok.

**Comment by jmazzitelli** [comment]
@mattmahoneyrh is putting his master build into OSSM 2.1 and see how it works there (which is the main reason why this needs to be fixed - so he's verifying this on OSSM 2.1).

**Pull request merged**
